packages/astro/test/astro-scoped-styles.test.js

```diff
@@ -22,7 +22,13 @@ ScopedStyles('Scopes rules correctly', () => {
     '.class :global(*)': `.class.${className} *`,
     '.class :global(.nav:not(.is-active))': `.class.${className} .nav:not(.is-active)`, // preserve nested parens
     '.class :global(ul li)': `.class.${className} ul li`, // allow doubly-scoped selectors
-    '.class:not(.is-active)': `.class.${className}:not(.is-active)`, // Note: the :not() selector can NOT contain multiple classes, so this is correct; if this causes issues for some people then it‘s worth a discussion
+    ':global(body:not(.is-light)).is-dark,:global(body:not(.is-dark)).is-light': `body:not(.is-light).is-dark,body:not(.is-dark).is-light`, // :global() can contain parens, and can be chained off of
+    ':global(.foo):global(.bar)': '.foo.bar', // more :global() shenanigans
+    '.class:global(.bar)': `.class.bar`, // this is technically a “useless“ :global() but it should still be extracted
+    '.class:not(.is-active):not(.is-disabled)': `.class.${className}:not(.is-active):not(.is-disabled)`, // Note: the :not() selector can NOT contain multiple classes, so this is correct; if this causes issues for some people then it‘s worth a discussion
+    ':hover.a:focus': `.${className}:hover.a:focus`, // weird but still valid (yes, it’s valid)
+    '*:hover': `.${className}:hover`,
+    ':not(.is-disabled).a': `.${className}:not(.is-disabled).a`, // also valid
```

Comment:
> LGTM, but bonus points if you can come up with more weird edge cases to test for in the test file :)

Done 🙂

Reply:
So for more context, there’s one really weird scenario where the final result is ambiguous:

```
:global(.bar).foo
```

Now, although logically this doesn’t make sense (if one class is global, but another class is scoped, they’d always be effectively scoped), semantically it’s correct and there are two possible ways to parse this:

1. **Scoped**: `.bar.foo.astro-XXXXXX`
2. **Global**: `.bar.foo` 

While #1 may be technically-correct (after all, part of the selector was unscoped), it’s actually deceptively hard to get right in every possible iteration of the syntax. So in favor of a much, much, much simpler solution, #2 was applied: if you use `:global()` on an element, we’ll just assume that you wanted that whole thing global. Even if you tacked weird stuff outside the parens.

This also seems to solve the spirit of #1155 as well—generated CSS behaving weirdly and resulting in broken styles.

So #2 is simpler and hopefully more predictable. Win–win!